src_c/rwobject.c

```diff
@@ -700,90 +662,83 @@ _pg_rw_read(SDL_RWops *context, void *ptr, size_t size, size_t maxnum)
-                    else {
-                        PyErr_Format(PyExc_FileNotFoundError, 
-                                     "No such file or directory: '%S'.", obj);
-                    }
-                    Py_DECREF(cwd);
-                    Py_DECREF(path);
-                    Py_DECREF(isabs);
-                }
-                else {
-                    PyErr_Format(PyExc_FileNotFoundError, 
-                                 "No such file or directory: '%S'.", obj);
-                }
-                return NULL;
+    SDL_RWops *rw = NULL;
+    PyObject *oencoded;
+    char* encoded = NULL;
+    if (!obj) {
+        // forward any errors
+        return NULL;
+    }
+
+    oencoded = pg_EncodeString(obj, "UTF-8", NULL, NULL);
+    if (!oencoded || oencoded == Py_None) {
+        /* if oencoded is NULL, we are forwarding an error. If it is None, the
+         * object passed was not a bytes/string/pathlib object so handling of
+         * that is done after this function, exit early here */
+        Py_XDECREF(oencoded);
+        return NULL;
+    }
+
+    encoded = Bytes_AS_STRING(oencoded);
+    Py_DECREF(oencoded);
+
+    rw = SDL_RWFromFile(encoded, "rb");
+    if (rw) {
+        /* adding the extension to the hidden data for RWops from files */
+        /* this is necessary to support loading functions that rely on
+            * file extensions in a convenient way. File-like objects use this
+            * field for a helper object. */
+        char *extension = NULL;
+        char *ext = strrchr(encoded, '.');       
+        if (ext && strlen(ext) > 1) {
+            ext++;
+            extension = malloc(strlen(ext)+1);
+            if (!extension) {
+                return (SDL_RWops *)PyErr_NoMemory();
             }
+            strcpy(extension, ext);
         }
-        SDL_ClearError();
+        rw->hidden.unknown.data1 = (void *)extension;
+        return rw;
     }
+
+    /* Clear SDL error and set our own error message for filenotfound errors
+     * TODO: Check SDL error here and forward any non filenotfound related 
+     * errors correctly here */
+    SDL_ClearError();
+
+    PyObject *cwd = NULL, *path = NULL, *isabs = NULL;
+    if (!os_module)
+        goto error;
+
+    cwd = PyObject_CallMethod(os_module, "getcwd", NULL);
+    if (!cwd)
+        goto error;
+
+    path = PyObject_GetAttrString(os_module, "path");
+    if (!path)
+        goto error;
+
+    isabs = PyObject_CallMethod(path, "isabs", "O", obj);
+    Py_DECREF(path);
+    if (!isabs)
+        goto error;
+
+    if (isabs == Py_True) 
+        goto error;
+
+    PyErr_Format(PyExc_FileNotFoundError,
+                        "No file '%S' found in working directory"
+                        " '%S'.", obj, cwd);                       
+
+    Py_DECREF(cwd);
+    Py_DECREF(isabs);
+    return NULL;
+
+error:
+    Py_XDECREF(cwd);
+    Py_XDECREF(isabs);
+    PyErr_Format(PyExc_FileNotFoundError, "No such file or directory: '%S'.", obj);
```

Comment:
It's a bit strange to have a state labeled error that is fine to go to in normal execution.

Maybe it should be named like `simple_case` or something?

Reply:
yeah, true